lib/core/stream.nit

```diff
@@ -95,32 +95,71 @@ end
 abstract class Reader
 	super Stream
 
+	# Read a byte directly from the underlying stream, without
+	# considering any eventual buffer
+	protected fun raw_read_byte: Int is abstract
+
+	# Read at most `max` bytes from the underlying stream into `buf`,
+	# without considering any eventual buffer
+	#
+	# Returns how many bytes were read
+	protected fun raw_read_bytes(buf: CString, max: Int): Int do
+		var rd = 0
+		for i in [0 .. max[ do
+			var b = raw_read_byte
+			if b < 0 then break
+			buf[i] = b.to_b
+			rd += 1
+		end
+		return rd
+	end
+
 	# Reads a character. Returns `null` on EOF or timeout
 	fun read_char: nullable Char is abstract
 
 	# Reads a byte. Returns a negative value on error
-	fun read_byte: Int is abstract
+	fun read_byte: Int do
+		var llen = lookahead_length
+		if llen == 0 then return raw_read_byte
+		var lk = lookahead
+		var b = lk[0].to_i
+		if llen == 1 then
+			lookahead_length = 0
+		else
+			lk.lshift(1, llen - 1, 1)
```

Comment:
ouch! Shouldn't a cursor better that shifting the whole lookahead buffer?

Reply:
Yeah, that would be better indeed. But, since the codec will need a sequence of bytes to determine if the chars to read are valid, I don't have much of a choice.
Besides, this should be a very rare case as it will only happen when an invalid char was read (which, if you chose your encoding properly should not happen that much), and the lookahead should not be too long (4 bytes max in the case of unicode).